library/src/main/java/com/novoda/downloadmanager/lib/DownloadProvider.java

```diff
@@ -408,10 +408,12 @@ private Uri insertDownload(Uri uri, ContentValues values, SQLiteDatabase db, int
                 dest = DownloadsDestination.DESTINATION_CACHE_PARTITION;
             }
             if (dest == DownloadsDestination.DESTINATION_FILE_URI) {
-                getContext().enforcePermission(
-                        android.Manifest.permission.WRITE_EXTERNAL_STORAGE, Binder.getCallingPid(), Binder.getCallingUid(),
-                        "need WRITE_EXTERNAL_STORAGE permission to use DESTINATION_FILE_URI");
-                checkFileUriDestination(values);
+                if (!getFileUriDestinationPath(values).startsWith(Environment.getDataDirectory() + "/")) {
```

Comment:
Hmm. Thanks, @rock3r  - I'm struggling with this, though, because at this point in the code this destination file path is nowhere near being turned into an actual File that you could call getAbsolutePath() on, and there's a whole lot of logic elsewhere, largely in the Helpers class, including the generateSaveFile(), getFullPath(), and chooseFileName() methods, which come into play before actually turning this into a File. I'd be reluctant to bring all that code in to play at this point. Also, as far as I can tell, the only way the COLUMN_FILE_NAME_HINT in the DB could be a relative URI is if you passed one in via Request.setDestinationUri() (rather than using one of the other setDestination... methods). If you did that, and you used a path that begins with the result of Environment.getDataDirectory()...well, then I think it would actually just work as an internal-storage download, and I'm not sure that's problematic. It's quite possible I'm all wet, though. This is from an hour or two of poking through the code.